feat: Set up docker for Endatix API and Endatix Hub

dspetrov opened the pull request:
# Set up docker for Endatix API and Endatix Hub

## Description
- Creates Dockerfile for Endatix.WebHost .NET project and Endatix Hub, and docker-compose that includes the creation of the images from the two Dockerfile, getting MS SQL Server image and configuring them.
- Endatix Hub had several issues during its build and they were fixed.
- The Next.js app is built using output "standalone" to reduce the size of the Docker image
- Healthchecks are added when running the containers
- `docker-compose.yml` gets configurations from environment variables. They should either be set in the environment or for dev and testing can be set in a file `.env` with contents like:
```
ASPNETCORE_ENVIRONMENT=Development
SECURITY_JWT_SIGNING_KEY=L2yGC_Vpd3k#L[<9Zb,h?.HT:n'T/5CTDmBpDskU?NAaT$sLfRU
SECURITY_KEY_DEV_USERS_0_EMAIL=developer@endatix.com
SECURITY_KEY_DEV_USERS_0_PASSWORD=password
SENDGRID_API_KEY=[api key]
SQLSERVER_SA_PASSWORD=[the desired password for the sql server container]
```
- Images are created and uploaded in Docker Hub:
  - https://hub.docker.com/repository/docker/endatix/endatix-api/general
  - https://hub.docker.com/repository/docker/endatix/endatix-hub/general
- A new folder `\docker` is created in the repo. It includes a file `docker-compose.dev.yml` that gets the images from Docker Hub and runs them locally (without building them). The folder also includes scripts to assist the process and a readme with details how to get the apps running.

## How to test

### Prerequisites
Docker needs to be installed for testing and commands docker and docker-compose are utilized. Installing Docker Desktop is sufficient to have all this.

### Test scenarios

#### Build and execution
- Ensure the env variables from the description are either set locally or present in a file .env in the root repo folder
- In the root repo folder run `docker-compose build`. In a while the images should be successfully created and set in the local Docker.
- Run `docker-compose up -d` to get the containers running
- Try http://localhost:5001/swagger - authenticate with "developer@endatix.com" / "password", use the token to get the forms and possibly try some other endpoints
- Try http://localhost:3000 - Clicking around should show some info messages how to learn more about Endatix

#### Execution of Docker Hub images
- Ensure there are no Endatix API and Hub images in the local Docker so getting them from Docker Hub can be tested
- Follow https://github.com/endatix/endatix/blob/feature/issue-84-set-up-docker/docker/README.md

## Related Issues
closes #84
closes #85

## Type of Change
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

orilux commented:
I encountered several issues and only managed to get this partially working.

Firstly I was not able to get Docker to build the containers due to a NextJS build error:

Error occurred prerendering page "/forms". Read more: https://nextjs.org/docs/messages/prerender-error
13.19
13.19 TypeError: fetch failed
13.19     at node:internal/deps/undici/undici:13185:13
13.19     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
13.19     at async t (/app/.next/server/app/forms/page.js:1:14655)
13.19     at async p (/app/.next/server/app/forms/page.js:1:7884)
13.19    Generating static pages (2/8)
13.19    Generating static pages (4/8)
13.19    Generating static pages (6/8)
13.19 TypeError: fetch failed
13.19     at node:internal/deps/undici/undici:13185:13
13.19     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
13.19     at async t (/app/.next/server/app/forms/page.js:1:14655)
13.19     at async p (/app/.next/server/app/forms/page.js:1:7884) {
13.19   digest: '1547679062',
13.19   [cause]: AggregateError [ECONNREFUSED]:
13.19       at internalConnectMultiple (node:net:1117:18)
13.19       at afterConnectMultiple (node:net:1684:7)
13.19       at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
13.19     code: 'ECONNREFUSED',
13.19     [errors]: [ [Error], [Error] ]
13.19   }
13.19 }
13.19 Failed to build /forms/page: /forms after 1 attempts.

It appears that Next was trying to prerender the Forms page, which requires a connection to the Endatix API. As advised, I tried to run the build while the API is running. Not surprisingly that did not resolve the problem, because I couldn't figure out how to provide the auth token to the endatix-hub app.

I decided to modify the endatix-hub app locally to hardcode the token so that I can resume with the PR review. I modified the `endatix-hub\services\api.ts` file to add an auth entry in the header:

`export const getForms = async (): Promise<Form[]> => {
  const response = await fetch(`${API_BASE_URL}/forms`, {
    headers: {
      Authorization: `Bearer [my token]`
    },
  });`
  
This finally got endatix-hub to build, which in turn got the Docker containers to build and run.

I initially had trouble opening both localhost:5001 and localhost:3000, until I realized they don't support HTTPS. After accessing them successfully over the non-secure protocol, I was only able to interact with the Authorization endpoint. Everything else returns 500 after a about 10 second delay, which suggests that the database cannot be accessed. I assume my connection string is not in the correct format, but I tried every variation I can think of without any luck.

dspetrov commented:
> I encountered several issues and only managed to get this partially working.
> 
> Firstly I was not able to get Docker to build the containers due to a NextJS build error:
> 
> ```
> Error occurred prerendering page "/forms". Read more: https://nextjs.org/docs/messages/prerender-error
> 13.19
> 13.19 TypeError: fetch failed
> 13.19     at node:internal/deps/undici/undici:13185:13
> 13.19     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
> 13.19     at async t (/app/.next/server/app/forms/page.js:1:14655)
> 13.19     at async p (/app/.next/server/app/forms/page.js:1:7884)
> 13.19    Generating static pages (2/8)
> 13.19    Generating static pages (4/8)
> 13.19    Generating static pages (6/8)
> 13.19 TypeError: fetch failed
> 13.19     at node:internal/deps/undici/undici:13185:13
> 13.19     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
> 13.19     at async t (/app/.next/server/app/forms/page.js:1:14655)
> 13.19     at async p (/app/.next/server/app/forms/page.js:1:7884) {
> 13.19   digest: '1547679062',
> 13.19   [cause]: AggregateError [ECONNREFUSED]:
> 13.19       at internalConnectMultiple (node:net:1117:18)
> 13.19       at afterConnectMultiple (node:net:1684:7)
> 13.19       at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
> 13.19     code: 'ECONNREFUSED',
> 13.19     [errors]: [ [Error], [Error] ]
> 13.19   }
> 13.19 }
> 13.19 Failed to build /forms/page: /forms after 1 attempts.
> ```
> 
> It appears that Next was trying to prerender the Forms page, which requires a connection to the Endatix API. As advised, I tried to run the build while the API is running. Not surprisingly that did not resolve the problem, because I couldn't figure out how to provide the auth token to the endatix-hub app.
> 
> I decided to modify the endatix-hub app locally to hardcode the token so that I can resume with the PR review. I modified the `endatix-hub\services\api.ts` file to add an auth entry in the header, starting with line 5:
> 
> ```
> export const getForms = async (): Promise<Form[]> => {
>   const response = await fetch(`${API_BASE_URL}/forms`, {
>     headers: {
>       Authorization: `Bearer [my token]`
>     },
>   });
> ```
> 
> This finally got endatix-hub to build, which in turn got the Docker containers to build and run.
> 
> I initially had trouble opening both localhost:5001 and localhost:3000, until I realized they don't support HTTPS. After accessing them successfully over the non-secure protocol, I was only able to interact with the Authorization endpoint. Everything else returns 500 after a about 10 second delay, which suggests that the database cannot be accessed. I assume my connection string is not in the correct format, but I tried every variation I can think of without any luck.

@orilux, I've done some changes so you should not have the described issues anymore. I also updated the description of the PR so please read the updated one. There are two test scenarios now, described there.

Running HTTPS inside the containers needs additional changes and I will create another issue for it.

PeterMarinov commented:
My feedback for each scenario:

## Scenario 1: Build and execution

**Step 1**: Command `docker-compose build`
*  Command `docker-compose build` is not recognized as command on Mac => Instead we should use `docker compose build` on mac. This is the v2 of Docker compose and is recommended - https://docs.docker.com/compose/migrate/#what-are-the-differences-between-compose-v1-and-compose-v2. `docker compose build` works ✅ 
* Recommendation: why not adding `.env` file by default in the codebase with default params that can be set? This will reduce manual steps. Alternatively add CLI tool that generates the .env file with a command

**Step 2**: `docker-compose up -d` - let's add to the description that `-d` means to run the containers in detached mode in the background. Command again should be compose v2 `docker compose up -d`
* Had issues with env variables not found below. I found that the issue was with the variables in the PR description (e.g. SECURITY_KEY_DEV_USERS_0_EMAIL=developer@endatix.com & SECURITY_KEY_DEV_USERS_0_PASSWORD=password). They are correct in the Docker Readme.
```bash
WARN[0000] The "sLfRU" variable is not set. Defaulting to a blank string. 
WARN[0000] The "SECURITY_DEV_USERS_0_PASSWORD" variable is not set. Defaulting to a blank string. 
WARN[0000] The "SECURITY_DEV_USERS_0_EMAIL" variable is not set. Defaulting to a blank string.
```
On the `WARN[0000] The "sLfRU" variable is not set. Defaulting to a blank string.` error it comes from the last 4 digits of the `SECURITY_JWT_SIGNING_KEY=L2yGC_Vpd3k#L[<9Zb,h?.HT:n'T/5CTDmBpDskU?NAaT$sLfRU` key provided in the PR description. Adding "" doesn't solve the issue for me. Issue is that bash uses the "$" to mark the definition of new variable - e.g. `$sLfRU` is interpreted as defining `sLfRU` 
* Execution Result - `Error response from daemon: driver failed programming external connectivity on endpoint endatix-endatix-db-1 (236f353a6094fd22e0d16085a12e38ad060e173c74141217f3e4c6c0d9002b2a): Bind for 0.0.0.0:1433 failed: port is already allocated`. - seems that 1433 was taken as port for me. Maybe consider using less popular port as many people might have a DB using it already
* After clearing my port, I have encountered this issue
```
✘ Container endatix-endatix-db-1                                                                                                                            Error25.7s                                                                                 
 ! endatix-db The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s 
```
Issue is with my M1 Mac chip. I spent a lot of time to troubleshoot. At the end I fixed with adding `platform: linux/amd64/v8` for the "endatix-db:" to the docker-compose.yml in the root. However, this is a hack as it only adds support for v1 of ARM chips, e.g. M1. 
**Recommendation** - add muti-platform image support with buildx - https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

After that it worked - ✅
Default Urls from docker were not working - we should add support for https://localhost:5001 to point to somewhere or redirect to swagger

## Scenario 2: Execution of Docker Hub images
### Step 1: Perfect to use the `curl` - ✅ .Maybe add the text that this should be done in empty folder, e.g. doesn't have to be in the source code. The context of the Readme is the source code, where the files are already downloaded so it's a bit counter-intuitive to download them again, I think

### Step 2: Run the docker compose
* I had the issue of `Docker Compose is not installed. Please install Docker Compose before proceeding.` coming from the fact I am using Compose v2 - https://docs.docker.com/compose/migrate/. The line below is causing it
```bash
# Check if Docker Compose is installed
if ! [ -x "$(command -v docker-compose)" ]; then
  echo "Docker Compose is not installed. Please install Docker Compose before proceeding."
  exit 1
fi
```
Recommendation - migrate setup script to use v2 - https://stackoverflow.com/questions/72928891/how-can-i-check-if-docker-compose-plugin-is-installed

After I have removed this line and changed to v2 e.g. changing `docker-compose` to `docker compose` it run. Few issues I see:
* `Enter ASPNETCORE_ENVIRONMENT (e.g., Development): Development` - the default should be Development. Will be less clicks and issues. Alternatively we can set multiple choice + other. Not sure if we can do this with CLI
* `Enter SECURITY_JWT_SIGNING_KEY (JWT signing key to be used in Endatix API):` - No guidance on its length or symbols. Also this will also cause issues if users type "$" or other symbol not liked by .env files.  We should instead auto-generate this without asking the user anything
* `Enter SECURITY_DEV_USERS_0_EMAIL (Test user email for API auth): ` - no validation on this one. I can enter anything. We should move this away in CLI tool or add this as UI step in the Endatix-Hub where we can have better UI to set the user
* `SECURITY_DEV_USERS_0_PASSWORD (Test user password for API auth, min 8 chars):` - same here. I can enter 2 symbols
* `Enter SQLSERVER_SA_PASSWORD (SQL Server container admin user password, min 8 chars, include uppercase, lowercase, digit and special char):` - no validation here as well. We should perhaps use other ways to generate it e.g. CLI tool 
* Everytime we start the setup the .env file is deleted and I needed to enter the values over and over again. Imo a better Dev X will be to detect values and allow to skip changing them one at a time or all at once.
* It run ✅  for me once the changes for docker compose v2 were added + the "platform: linux/amd64/v8" for the "endatix-db"
* Recommendation: add few lines to echo the Urls at the end of the docker command. Now it shows the following line:
`Docker Compose started successfully. Your containers should be up and running.`
We should add the Urls and some successfully formatted message

dspetrov commented:
@PeterMarinov, thanks for the detailed feedback.
I've made changes according to some of the points and additionally added default values in the scripts and more details in the readme.
I agree with most of the comments that don't lead to changes now (and myself considered some of them) and they will be great enhancements for the future.
A couple of notes:
- Actually `docker-compose` is an alias for `docker compose` so it was running v2 even before the changes. Of course, having it explicitly execute `docker-compose` is a better approach.
- I also thought about committing the .env file but this is not a good practice because sooner or later someone could commit a sensitive piece of data.

PeterMarinov commented on docker/setup.bat:
```diff
@@ -0,0 +1,94 @@
+IF "%ASPNETCORE_ENVIRONMENT%"=="" set "ASPNETCORE_ENVIRONMENT=Development"
+
+set /p SECURITY_JWT_SIGNING_KEY="Enter SECURITY_JWT_SIGNING_KEY (JWT signing key to be used in Endatix API, skip to use the default value): "
+IF "%SECURITY_JWT_SIGNING_KEY%"=="" set "SECURITY_JWT_SIGNING_KEY=L2yGC_Vpd3k#L[<9Zb,h?.HT:n'T/5CTDmBpDskU?NAaT$sLfRU"
```
@dspetrov  - it's much better, intuitive and faster now - definitely makes a difference!

Only one issue on my Mac I have this error:

`failed to read ../../Code/endatix/endatix/docker-test/.env: line 3: unexpected character "/" in variable name "T/5CTDmBpDskU?NAaT$sLfRU'"`

Can you remove the / from the default key. It work for me with it removed?

PeterMarinov commented on docker/setup.sh:
```diff
@@ -0,0 +1,74 @@
+  exit 1
+fi
+
+echo "Docker Compose started successfully. Your containers should be up and running. Access the platform at:"
```
This is much better now. I really like it. I want to propose another output

<img width="750" alt="image" src="https://github.com/user-attachments/assets/69ad1390-d4a4-4998-8179-4cbb56f1187a">

Here is the bash code. Only tested this on my Mac:

```bash
echo "Docker Compose started successfully! 🚀 "
echo "---------------------------------------------------"
echo "Access the Endatix containers at:"
echo "          🔗 http://localhost:5001/swagger for Endatix API"
echo "          🔗 http://localhost:3000 for Endatix Hub"
echo ""
echo "✅ The initial setup of Endatix Platform is completed."
echo "---------------------------------------------------"
echo "To stop the containers use \`docker compose -f docker-compose.dev.yml stop\`."
echo "To start again the containers use \`docker compose -f docker-compose.dev.yml start\`."
echo "To delete the containers use \`docker compose -f docker-compose.dev.yml down\`."
echo ""
```

PeterMarinov reviewed (changes requested):
@dspetrov  - this is great now. Makes huge difference. I have requested just 2 changes

PeterMarinov reviewed (approved):
Approved. Good to go
